docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,541 @@
+        params, grads)
+    return new_params, MomentumState(new_momentum)
+
+  return Optimizer(init_fn, update_fn)
+```
+
+Note that unlike in the [previous API], the function `update_fn()` will be
+called directly in the train loop and must take care of processing the pytrees
+of parameters and optimizer state with a `jax.tree_multimap()`. For more
+complex optimizers this can become a hassle but we can provide a small
+functional helper `@multimap` that takes care of this:
+
+```python
+  def update_fn(
+    grads: Gradients, params: Params, state: OptimizerState
+  ) -> Tuple[Params, OptimizerState]:
+
+    @multimap
+    def inner(param, grad, momentum):
+      new_momentum = momentum * beta + grad
+      new_param = param - learning_rate * grad
+      return new_param, new_momentum
+
+    new_params, new_momentum = inner(params, grads, state.momentum)
+    return new_params, MomentumState(new_momentum)
+```
+
+Remarks:
+
+- We'll call `Optimizer` instances from the functional API `optim`, and refer
+  to `flax.optim.Optimizer` instances from the [previous API] as `optimizer`
+  for the remainder of this document.
+- Compared with the [previous API], this new API does not have the option to
+  update hyper parameters. With Optax, updating the hyper parameters is instead
+  done by defining a schedule that takes care of updating the hyper parameters
+  during training (see the [using Optax] code snippet for an example).
+- Using this simple API directly makes the user code a bit more verbose than
+  using the [previous API], but the difference is smaller with Linen. If you
+  care about a boilerplate-free training loop see the proposed [Linen helper]
+  further down.
+
+## Using Optax
+[Using Optax]: #using-optax
+
+Optax is centered on the idea of composable gradient transformations. Since
+above-mentioned [functional API] is based on Optax's interface it is
+straightforward to wrap an Optax transformation to comply with the `Optimizer`
+interface:
+
+```python
+import optax
+
+
+def optax_optimizer(tx: optax.GradientTransformation) -> Optimizer:
+  """Wraps an Optax transformation to update parameters directly."""
+
+  def update_fn(
+    grads: Gradients, params: Params, state: OptimizerState
+  ) -> Tuple[Params, OptimizerState]:
+    updates, state = tx.update(grads, state, params)
+    params = optax.apply_updates(params, updates)
+    return params, state
+
+  return Optimizer(tx.init, update_fn)
+
+tx = optax.chain(
+    optax.trace(decay=0.9, nesterov=False),
+    optax.scale_by_schedule(lambda step: -get_learning_rate(step))
+)
+
+optim = optax_optimizer(tx)
+```
+
+Remarks:
+
+- The only difference is that the Optax udpate function returns a transformed
+  gradient and the updated optimizer state (which makes them composable), so we
+  apply this to the parameters in a final step and return the updated parameters
+  and optimizer state.
+- We can use a `get_learning_rate()` that returns the learning rate depending on
```

Comment:
we often log the learning rate in metrics. How do I retrieve metrics like learning rate and grad norm (when clipping is used) from the update?

Reply:
Mentioning this in the discussion under `Optax Training Step`.

Reply:
The current proposal for scheduling hyperparameters in Optax stores the hyperparameters in the `OptState`, which makes logging them convenient. We'd appreciate feedback on this design on the [PR](https://github.com/deepmind/optax/pull/48)!